In [1]:
from bs4 import BeautifulSoup
import pandas
import urllib
import csv
import re
import urllib
import bs4

from urllib import request


# Scrapping

In [2]:
df_All_Book=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Data_Traitee/Books_v4.csv",sep=";",index_col=None)

In [3]:
df_Book_needed=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Data_Traitee/Books_6508.csv",sep=",",index_col=None)
df_Book_needed=df_Book_needed[["ISBN"]]

In [4]:
print(df_All_Book.shape)
print("-----------------------")
print(df_All_Book.head())
print("--------------------------------------------------------------------------------------------")

print(df_Book_needed.shape)
print("-----------------------")
print(df_Book_needed.head())
print("--------------------------------------------------------------------------------------------")


(142026, 9)
-----------------------
   Unnamed: 0       ISBN                                         Book.Title  \
0           2    2005018                                       Clara Callan   
1           3   60973129                               Decision in Normandy   
2           4  374157065  Flu: The Story of the Great Influenza Pandemic...   
3           6  399135782                             The Kitchen God's Wife   
4           7  425176428  What If?: The World's Foremost Military Histor...   

            Book.Author  Year.Of.Publication                 Publisher  \
0  Richard Bruce Wright                 2001     HarperFlamingo Canada   
1          Carlo D'Este                 1991           HarperPerennial   
2      Gina Bari Kolata                 1999      Farrar Straus Giroux   
3               Amy Tan                 1991          Putnam Pub Group   
4         Robert Cowley                 2000  Berkley Publishing Group   

                                         Ima

In [5]:
df_Final=pandas.merge(df_Book_needed, df_All_Book, on=['ISBN'])

In [6]:
print(df_Final.shape)
df_Final.head()

(6508, 9)


ISBN  Unnamed: 0            Book.Title      Book.Author  \
0  034545104X        2967  Flesh Tones: A Novel       M. J. Rose   
1   446520802       11055          The Notebook  Nicholas Sparks   
2  038550120X        9296       A Painted House     JOHN GRISHAM   
3   425115801        2031             Lightning   Dean R. Koontz   
4   449006522         228   Manhattan Hunt Club        JOHN SAUL   

   Year.Of.Publication                 Publisher  \
0                 2002          Ballantine Books   
1                 1996              Warner Books   
2                 2001                 Doubleday   
3                 1996  Berkley Publishing Group   
4                 2002          Ballantine Books   

                                         Image.URL.S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/0446520802.0...   
2  http://images.amazon.com/images/P/038550120X.0...   
3  http://images.amazon.com/images/P/0425115801.0...   
4  http://images.amazon.com/images/P/0449006522.0...   

                                         Image.URL.M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/0446520802.0...   
2  http://images.amazon.com/images/P/038550120X.0...   
3  http://images.amazon.com/images/P/0425115801.0...   
4  http://images.amazon.com/images/P/0449006522.0...   

                                         Image.URL.L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/0446520802.0...  
2  http://images.amazon.com/images/P/038550120X.0...  
3  http://images.amazon.com/images/P/0425115801.0...  
4  http://images.amazon.com/images/P/0449006522.0...

### Tous les livres

In [7]:
df_Final.shape

(6508, 9)

Comme il y a 6508 livre, on divise la base en 7

In [8]:
Liste_df=[df_Final[0:1000],df_Final[1000:2000],df_Final[2000:3000],df_Final[3000:4000],df_Final[4000:5000],df_Final[5000:6000],df_Final[6000:]]
Liste_book=[0,1,0,0,0,0,0]
Liste_df_book0=list()
Liste_df_book1=list()
Liste_df_book3=list()

#### Livre 1 à 1000

In [ ]:
Z=0
df=Liste_df[Z]
l=[]
l2=[]
book=Liste_book[Z]
for i in range(df.shape[0]):
    book=book+1
    l.append(df["ISBN"][df.index[i]])
    l2.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN"][df.index[i]])
    if(len(ISBN)<10):
        while(len(ISBN)<10):
            ISBN ='0'+ISBN
    url = "https://www.bookfinder.com/search/?author=&title=&lang=en&isbn=" + str(ISBN) +"&new_used=*&destination=fr&currency=EUR&mode=basic&st=sr&ac=qr"         
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
           
    h=''
    for link in soup.find_all("span", {'itemprop' : 'author'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
  
    h=''
    for link in soup.find_all("span", {'itemprop' : 'publisher'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
    
    h=''
    for link in soup.find_all("link", {'itemprop' : 'bookformat'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')

    
    h=''
    for link in soup.find_all("span", {'itemprop' : 'inLanguage'}):
            h=link.get_text()
            h=h.replace('Language_','')
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
        
    k=0
    for link in soup.find("h1"):
        h=''
        if(k/2 == 0):
            l2.append(str(link)[:13])
        else :
            h=str(link)
            h=h.replace('<span itemprop="isbn">','')
            h=h.replace('</span>','')
            l2.append(h)
        k=k+1        
        
    h=''    
    for link in soup.find_all("span", {'class' : 'book-rating-average text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE') 
        
        
    h=''
    for link in soup.find_all("span", {'class' : 'book-rating-provider text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE')
        
            
Liste_book.append(book)

            
m=[]
m = [ [l[j], l[j+1], l[j+2],l[j+3],l[j+4]]  for j in range(0, int(len(l)), 5) ]
df_book0 = pandas.DataFrame(index = range(len(m)), columns = ["ISBN","AUTEUR", "PUBLISHER","EDITION","LANGUAGE"])
for i in range(len(m)):
    for variable in range(len(df_book0.columns)):
        df_book0.at[i, df_book0.columns[variable]] = m[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_1_1000..csv"
df_book0.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book0.append(df_book0)

m2=[]
m2 = [ [l2[j], l2[j+1], l2[j+2],l2[j+3],l2[j+4]]  for j in range(0, int(len(l2)), 5) ]
df_book1 = pandas.DataFrame(index = range(len(m2)), columns =["ISBN","ISBN13","ISBN10", "Avg_Rating","Nb_Ratings"])
for i in range(len(m2)):
    for variable in range(len(df_book1.columns)):
        df_book1.at[i, df_book1.columns[variable]] = m2[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_1_1000.csv"
df_book1.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book1.append(df_book1)


In [ ]:
book

In [ ]:
df=df_book1
l=[]



for i in range(df.shape[0]):
    l.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN13"][df.index[i]])
    if(len(ISBN)<13):
        while(len(ISBN)<13):
            ISBN ='0'+ISBN
    url = "https://www.bookdepository.com/book/" + str(ISBN) 
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
    
    add=''
    for link in soup.find_all('span'):
        if link.get("itemprop")=="numberOfPages":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
    add=''    
    for link in soup.find_all('span'):
        if link.get("itemprop")=="datePublished":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        
    add=''
    for link in soup.findAll("div", {'class' : 'item-tools'}):
        for t in link.findAll("span", {'class' : 'sale-price'}):
            add=t.get_text()
    for link in soup.findAll("p", {'class' : 'list-price'}):
            h=link.get_text()
            h=h.replace('\n','')
            h=h.replace('                                            List price: ','')  
            if(add==''):
                add=h
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        

n = [ l[j:j+4]  for j in range(0, int(len(l)), 4) ]
df_book3 = pandas.DataFrame(index = range(len(n)),columns =["ISBN","numberOfPages","datePublished","Price"])
for i in range(len(n)):
    for variable in range(4):
        df_book3.at[i,df_book3.columns[variable]] = n[i][variable]
Liste_df_book3.append(df_book3)

path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_1_1000.csv"
df_book3.to_csv(path, sep=";", encoding="utf-8")



#### Livre 1000 à 2000

In [9]:
Z=1
df=Liste_df[Z]
l=[]
l2=[]
book=Liste_book[Z]
for i in range(df.shape[0]):
    book=book+1
    l.append(df["ISBN"][df.index[i]])
    l2.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN"][df.index[i]])
    if(len(ISBN)<10):
        while(len(ISBN)<10):
            ISBN ='0'+ISBN
    url = "https://www.bookfinder.com/search/?author=&title=&lang=en&isbn=" + str(ISBN) +"&new_used=*&destination=fr&currency=EUR&mode=basic&st=sr&ac=qr"         
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
           
    h=''
    for link in soup.find_all("span", {'itemprop' : 'author'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
  
    h=''
    for link in soup.find_all("span", {'itemprop' : 'publisher'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
    
    h=''
    for link in soup.find_all("link", {'itemprop' : 'bookformat'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')

    
    h=''
    for link in soup.find_all("span", {'itemprop' : 'inLanguage'}):
            h=link.get_text()
            h=h.replace('Language_','')
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
        
    k=0
    for link in soup.find("h1"):
        h=''
        if(k/2 == 0):
            l2.append(str(link)[:13])
        else :
            h=str(link)
            h=h.replace('<span itemprop="isbn">','')
            h=h.replace('</span>','')
            l2.append(h)
        k=k+1        
        
    h=''    
    for link in soup.find_all("span", {'class' : 'book-rating-average text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE') 
        
        
    h=''
    for link in soup.find_all("span", {'class' : 'book-rating-provider text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE')
        
            
Liste_book.append(book)

            
m=[]
m = [ [l[j], l[j+1], l[j+2],l[j+3],l[j+4]]  for j in range(0, int(len(l)), 5) ]
df_book0 = pandas.DataFrame(index = range(len(m)), columns = ["ISBN","AUTEUR", "PUBLISHER","EDITION","LANGUAGE"])
for i in range(len(m)):
    for variable in range(len(df_book0.columns)):
        df_book0.at[i, df_book0.columns[variable]] = m[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_2_2000.csv"
df_book0.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book0.append(df_book0)

m2=[]
m2 = [ [l2[j], l2[j+1], l2[j+2],l2[j+3],l2[j+4]]  for j in range(0, int(len(l2)), 5) ]
df_book1 = pandas.DataFrame(index = range(len(m2)), columns =["ISBN","ISBN13","ISBN10", "Avg_Rating","Nb_Ratings"])
for i in range(len(m2)):
    for variable in range(len(df_book1.columns)):
        df_book1.at[i, df_book1.columns[variable]] = m2[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_2_2000.csv"
df_book1.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book1.append(df_book1)


Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.
Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.


In [10]:
df=df_book1
l=[]



for i in range(df.shape[0]):
    l.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN13"][df.index[i]])
    if(len(ISBN)<13):
        while(len(ISBN)<13):
            ISBN ='0'+ISBN
    url = "https://www.bookdepository.com/book/" + str(ISBN) 
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
    
    add=''
    for link in soup.find_all('span'):
        if link.get("itemprop")=="numberOfPages":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
    add=''    
    for link in soup.find_all('span'):
        if link.get("itemprop")=="datePublished":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        
    add=''
    for link in soup.findAll("div", {'class' : 'item-tools'}):
        for t in link.findAll("span", {'class' : 'sale-price'}):
            add=t.get_text()
    for link in soup.findAll("p", {'class' : 'list-price'}):
            h=link.get_text()
            h=h.replace('\n','')
            h=h.replace('                                            List price: ','')  
            if(add==''):
                add=h
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        

n = [ l[j:j+4]  for j in range(0, int(len(l)), 4) ]
df_book3 = pandas.DataFrame(index = range(len(n)),columns =["ISBN","numberOfPages","datePublished","Price"])
for i in range(len(n)):
    for variable in range(4):
        df_book3.at[i,df_book3.columns[variable]] = n[i][variable]
Liste_df_book3.append(df_book3)

path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_2_2000.csv"
df_book3.to_csv(path, sep=";", encoding="utf-8")



#### Livre 2000 à 3000

In [17]:
Z=2
df=Liste_df[Z]
l=[]
l2=[]
book=Liste_book[Z]
for i in range(df.shape[0]):
    book=book+1
    l.append(df["ISBN"][df.index[i]])
    l2.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN"][df.index[i]])
    if(len(ISBN)<10):
        while(len(ISBN)<10):
            ISBN ='0'+ISBN
    url = "https://www.bookfinder.com/search/?author=&title=&lang=en&isbn=" + str(ISBN) +"&new_used=*&destination=fr&currency=EUR&mode=basic&st=sr&ac=qr"         
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
           
    h=''
    for link in soup.find_all("span", {'itemprop' : 'author'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
  
    h=''
    for link in soup.find_all("span", {'itemprop' : 'publisher'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
    
    h=''
    for link in soup.find_all("link", {'itemprop' : 'bookformat'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')

    
    h=''
    for link in soup.find_all("span", {'itemprop' : 'inLanguage'}):
            h=link.get_text()
            h=h.replace('Language_','')
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
        
    k=0
    for link in soup.find("h1"):
        h=''
        if(k/2 == 0):
            l2.append(str(link)[:13])
        else :
            h=str(link)
            h=h.replace('<span itemprop="isbn">','')
            h=h.replace('</span>','')
            l2.append(h)
        k=k+1        
        
    h=''    
    for link in soup.find_all("span", {'class' : 'book-rating-average text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE') 
        
        
    h=''
    for link in soup.find_all("span", {'class' : 'book-rating-provider text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE')
        
            
Liste_book.append(book)

            
m=[]
m = [ [l[j], l[j+1], l[j+2],l[j+3],l[j+4]]  for j in range(0, int(len(l)), 5) ]
df_book0 = pandas.DataFrame(index = range(len(m)), columns = ["ISBN","AUTEUR", "PUBLISHER","EDITION","LANGUAGE"])
for i in range(len(m)):
    for variable in range(len(df_book0.columns)):
        df_book0.at[i, df_book0.columns[variable]] = m[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_3_3000.csv"
df_book0.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book0.append(df_book0)

m2=[]
m2 = [ [l2[j], l2[j+1], l2[j+2],l2[j+3],l2[j+4]]  for j in range(0, int(len(l2)), 5) ]
df_book1 = pandas.DataFrame(index = range(len(m2)), columns =["ISBN","ISBN13","ISBN10", "Avg_Rating","Nb_Ratings"])
for i in range(len(m2)):
    for variable in range(len(df_book1.columns)):
        df_book1.at[i, df_book1.columns[variable]] = m2[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_3_3000.csv"
df_book1.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book1.append(df_book1)


Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.
Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.
Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.
Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.
Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.
Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER.


IncompleteRead: IncompleteRead(133437 bytes read)

In [ ]:
book

In [ ]:
df=df_book1
l=[]



for i in range(df.shape[0]):
    l.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN13"][df.index[i]])
    if(len(ISBN)<13):
        while(len(ISBN)<13):
            ISBN ='0'+ISBN
    url = "https://www.bookdepository.com/book/" + str(ISBN) 
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
    
    add=''
    for link in soup.find_all('span'):
        if link.get("itemprop")=="numberOfPages":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
    add=''    
    for link in soup.find_all('span'):
        if link.get("itemprop")=="datePublished":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        
    add=''
    for link in soup.findAll("div", {'class' : 'item-tools'}):
        for t in link.findAll("span", {'class' : 'sale-price'}):
            add=t.get_text()
    for link in soup.findAll("p", {'class' : 'list-price'}):
            h=link.get_text()
            h=h.replace('\n','')
            h=h.replace('                                            List price: ','')  
            if(add==''):
                add=h
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        

n = [ l[j:j+4]  for j in range(0, int(len(l)), 4) ]
df_book3 = pandas.DataFrame(index = range(len(n)),columns =["ISBN","numberOfPages","datePublished","Price"])
for i in range(len(n)):
    for variable in range(4):
        df_book3.at[i,df_book3.columns[variable]] = n[i][variable]
Liste_df_book3.append(df_book3)

path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_3_3000.csv"
df_book3.to_csv(path, sep=";", encoding="utf-8")



#### Livre 3000 à 4000

In [ ]:
Z=3
df=Liste_df[Z]
l=[]
l2=[]
book=Liste_book[Z]
for i in range(df.shape[0]):
    book=book+1
    l.append(df["ISBN"][df.index[i]])
    l2.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN"][df.index[i]])
    if(len(ISBN)<10):
        while(len(ISBN)<10):
            ISBN ='0'+ISBN
    url = "https://www.bookfinder.com/search/?author=&title=&lang=en&isbn=" + str(ISBN) +"&new_used=*&destination=fr&currency=EUR&mode=basic&st=sr&ac=qr"         
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
           
    h=''
    for link in soup.find_all("span", {'itemprop' : 'author'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
  
    h=''
    for link in soup.find_all("span", {'itemprop' : 'publisher'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
    
    h=''
    for link in soup.find_all("link", {'itemprop' : 'bookformat'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')

    
    h=''
    for link in soup.find_all("span", {'itemprop' : 'inLanguage'}):
            h=link.get_text()
            h=h.replace('Language_','')
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
        
    k=0
    for link in soup.find("h1"):
        h=''
        if(k/2 == 0):
            l2.append(str(link)[:13])
        else :
            h=str(link)
            h=h.replace('<span itemprop="isbn">','')
            h=h.replace('</span>','')
            l2.append(h)
        k=k+1        
        
    h=''    
    for link in soup.find_all("span", {'class' : 'book-rating-average text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE') 
        
        
    h=''
    for link in soup.find_all("span", {'class' : 'book-rating-provider text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE')
        
            
Liste_book.append(book)

            
m=[]
m = [ [l[j], l[j+1], l[j+2],l[j+3],l[j+4]]  for j in range(0, int(len(l)), 5) ]
df_book0 = pandas.DataFrame(index = range(len(m)), columns = ["ISBN","AUTEUR", "PUBLISHER","EDITION","LANGUAGE"])
for i in range(len(m)):
    for variable in range(len(df_book0.columns)):
        df_book0.at[i, df_book0.columns[variable]] = m[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_4_4000.csv"
df_book0.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book0.append(df_book0)

m2=[]
m2 = [ [l2[j], l2[j+1], l2[j+2],l2[j+3],l2[j+4]]  for j in range(0, int(len(l2)), 5) ]
df_book1 = pandas.DataFrame(index = range(len(m2)), columns =["ISBN","ISBN13","ISBN10", "Avg_Rating","Nb_Ratings"])
for i in range(len(m2)):
    for variable in range(len(df_book1.columns)):
        df_book1.at[i, df_book1.columns[variable]] = m2[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_4_4000.csv"
df_book1.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book1.append(df_book1)


In [ ]:
df=df_book1
l=[]



for i in range(df.shape[0]):
    l.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN13"][df.index[i]])
    if(len(ISBN)<13):
        while(len(ISBN)<13):
            ISBN ='0'+ISBN
    url = "https://www.bookdepository.com/book/" + str(ISBN) 
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
    
    add=''
    for link in soup.find_all('span'):
        if link.get("itemprop")=="numberOfPages":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
    add=''    
    for link in soup.find_all('span'):
        if link.get("itemprop")=="datePublished":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        
    add=''
    for link in soup.findAll("div", {'class' : 'item-tools'}):
        for t in link.findAll("span", {'class' : 'sale-price'}):
            add=t.get_text()
    for link in soup.findAll("p", {'class' : 'list-price'}):
            h=link.get_text()
            h=h.replace('\n','')
            h=h.replace('                                            List price: ','')  
            if(add==''):
                add=h
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        

n = [ l[j:j+4]  for j in range(0, int(len(l)), 4) ]
df_book3 = pandas.DataFrame(index = range(len(n)),columns =["ISBN","numberOfPages","datePublished","Price"])
for i in range(len(n)):
    for variable in range(4):
        df_book3.at[i,df_book3.columns[variable]] = n[i][variable]
Liste_df_book3.append(df_book3)

path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_4_4000.csv"
df_book3.to_csv(path, sep=";", encoding="utf-8")



#### Livre 4000 à 5000

In [ ]:

Z=4
df=Liste_df[Z]
l=[]
l2=[]
book=0
for i in range(df.shape[0]):
    book=book+1
    l.append(df["ISBN"][df.index[i]])
    l2.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN"][df.index[i]])
    if(len(ISBN)<10):
        while(len(ISBN)<10):
            ISBN ='0'+ISBN
    url = "https://www.bookfinder.com/search/?author=&title=&lang=en&isbn=" + str(ISBN) +"&new_used=*&destination=fr&currency=EUR&mode=basic&st=sr&ac=qr"         
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
           
    h=''
    for link in soup.find_all("span", {'itemprop' : 'author'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
  
    h=''
    for link in soup.find_all("span", {'itemprop' : 'publisher'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
    
    h=''
    for link in soup.find_all("link", {'itemprop' : 'bookformat'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')

    
    h=''
    for link in soup.find_all("span", {'itemprop' : 'inLanguage'}):
            h=link.get_text()
            h=h.replace('Language_','')
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
        
    k=0
    for link in soup.find("h1"):
        h=''
        if(k/2 == 0):
            l2.append(str(link)[:13])
        else :
            h=str(link)
            h=h.replace('<span itemprop="isbn">','')
            h=h.replace('</span>','')
            l2.append(h)
        k=k+1        
        
    h=''    
    for link in soup.find_all("span", {'class' : 'book-rating-average text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE') 
        
        
    h=''
    for link in soup.find_all("span", {'class' : 'book-rating-provider text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE')
        
            
Liste_book[Z]=book

            
m=[]
m = [ [l[j], l[j+1], l[j+2],l[j+3],l[j+4]]  for j in range(0, int(len(l)), 5) ]
df_book0 = pandas.DataFrame(index = range(len(m)), columns = ["ISBN","AUTEUR", "PUBLISHER","EDITION","LANGUAGE"])
for i in range(len(m)):
    for variable in range(len(df_book0.columns)):
        df_book0.at[i, df_book0.columns[variable]] = m[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_5_5000.csv"
df_book0.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book0.append(df_book0)

m2=[]
m2 = [ [l2[j], l2[j+1], l2[j+2],l2[j+3],l2[j+4]]  for j in range(0, int(len(l2)), 5) ]
df_book1 = pandas.DataFrame(index = range(len(m2)), columns =["ISBN","ISBN13","ISBN10", "Avg_Rating","Nb_Ratings"])
for i in range(len(m2)):
    for variable in range(len(df_book1.columns)):
        df_book1.at[i, df_book1.columns[variable]] = m2[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_5_5000.csv"
df_book1.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book1.append(df_book1)


In [ ]:
df_book1=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_5_5000.csv",sep=";",encoding="utf-8")

In [ ]:
df=df_book1
l=[]



for i in range(df.shape[0]):
    l.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN13"][df.index[i]])
    if(len(ISBN)<13):
        while(len(ISBN)<13):
            ISBN ='0'+ISBN
    url = "https://www.bookdepository.com/book/" + str(ISBN) 
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
    
    add=''
    for link in soup.find_all('span'):
        if link.get("itemprop")=="numberOfPages":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
    add=''    
    for link in soup.find_all('span'):
        if link.get("itemprop")=="datePublished":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        
    add=''
    for link in soup.findAll("div", {'class' : 'item-tools'}):
        for t in link.findAll("span", {'class' : 'sale-price'}):
            add=t.get_text()
    for link in soup.findAll("p", {'class' : 'list-price'}):
            h=link.get_text()
            h=h.replace('\n','')
            h=h.replace('                                            List price: ','')  
            if(add==''):
                add=h
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        

n = [ l[j:j+4]  for j in range(0, int(len(l)), 4) ]
df_book3 = pandas.DataFrame(index = range(len(n)),columns =["ISBN","numberOfPages","datePublished","Price"])
for i in range(len(n)):
    for variable in range(4):
        df_book3.at[i,df_book3.columns[variable]] = n[i][variable]
Liste_df_book3.append(df_book3)

path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_5_5000.csv"
df_book3.to_csv(path, sep=";", encoding="utf-8")



#### Livre 5000 à 6000

In [ ]:
Z=5
df=Liste_df[Z]
l=[]
l2=[]
book=Liste_book[Z]
for i in range(df.shape[0]):
    book=book+1
    l.append(df["ISBN"][df.index[i]])
    l2.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN"][df.index[i]])
    if(len(ISBN)<10):
        while(len(ISBN)<10):
            ISBN ='0'+ISBN
    url = "https://www.bookfinder.com/search/?author=&title=&lang=en&isbn=" + str(ISBN) +"&new_used=*&destination=fr&currency=EUR&mode=basic&st=sr&ac=qr"         
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
           
    h=''
    for link in soup.find_all("span", {'itemprop' : 'author'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
  
    h=''
    for link in soup.find_all("span", {'itemprop' : 'publisher'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
    
    h=''
    for link in soup.find_all("link", {'itemprop' : 'bookformat'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')

    
    h=''
    for link in soup.find_all("span", {'itemprop' : 'inLanguage'}):
            h=link.get_text()
            h=h.replace('Language_','')
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
        
    k=0
    for link in soup.find("h1"):
        h=''
        if(k/2 == 0):
            l2.append(str(link)[:13])
        else :
            h=str(link)
            h=h.replace('<span itemprop="isbn">','')
            h=h.replace('</span>','')
            l2.append(h)
        k=k+1        
        
    h=''    
    for link in soup.find_all("span", {'class' : 'book-rating-average text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE') 
        
        
    h=''
    for link in soup.find_all("span", {'class' : 'book-rating-provider text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE')
        
            
Liste_book.append(book)

            
m=[]
m = [ [l[j], l[j+1], l[j+2],l[j+3],l[j+4]]  for j in range(0, int(len(l)), 5) ]
df_book0 = pandas.DataFrame(index = range(len(m)), columns = ["ISBN","AUTEUR", "PUBLISHER","EDITION","LANGUAGE"])
for i in range(len(m)):
    for variable in range(len(df_book0.columns)):
        df_book0.at[i, df_book0.columns[variable]] = m[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_6_6000.csv"
df_book0.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book0.append(df_book0)

m2=[]
m2 = [ [l2[j], l2[j+1], l2[j+2],l2[j+3],l2[j+4]]  for j in range(0, int(len(l2)), 5) ]
df_book1 = pandas.DataFrame(index = range(len(m2)), columns =["ISBN","ISBN13","ISBN10", "Avg_Rating","Nb_Ratings"])
for i in range(len(m2)):
    for variable in range(len(df_book1.columns)):
        df_book1.at[i, df_book1.columns[variable]] = m2[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_6_6000.csv"
df_book1.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book1.append(df_book1)


In [ ]:
book

In [ ]:
df=df_book1
l=[]



for i in range(df.shape[0]):
    l.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN13"][df.index[i]])
    if(len(ISBN)<13):
        while(len(ISBN)<13):
            ISBN ='0'+ISBN
    url = "https://www.bookdepository.com/book/" + str(ISBN) 
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
    
    add=''
    for link in soup.find_all('span'):
        if link.get("itemprop")=="numberOfPages":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
    add=''    
    for link in soup.find_all('span'):
        if link.get("itemprop")=="datePublished":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        
    add=''
    for link in soup.findAll("div", {'class' : 'item-tools'}):
        for t in link.findAll("span", {'class' : 'sale-price'}):
            add=t.get_text()
    for link in soup.findAll("p", {'class' : 'list-price'}):
            h=link.get_text()
            h=h.replace('\n','')
            h=h.replace('                                            List price: ','')  
            if(add==''):
                add=h
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        

n = [ l[j:j+4]  for j in range(0, int(len(l)), 4) ]
df_book3 = pandas.DataFrame(index = range(len(n)),columns =["ISBN","numberOfPages","datePublished","Price"])
for i in range(len(n)):
    for variable in range(4):
        df_book3.at[i,df_book3.columns[variable]] = n[i][variable]
Liste_df_book3.append(df_book3)

path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_6_6000.csv"
df_book3.to_csv(path, sep=";", encoding="utf-8")



#### Livre 6000 à ++

In [ ]:
Z=6
df=Liste_df[Z]
l=[]
l2=[]
book=Liste_book[Z]
for i in range(df.shape[0]):
    book=book+1
    l.append(df["ISBN"][df.index[i]])
    l2.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN"][df.index[i]])
    if(len(ISBN)<10):
        while(len(ISBN)<10):
            ISBN ='0'+ISBN
    url = "https://www.bookfinder.com/search/?author=&title=&lang=en&isbn=" + str(ISBN) +"&new_used=*&destination=fr&currency=EUR&mode=basic&st=sr&ac=qr"         
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
           
    h=''
    for link in soup.find_all("span", {'itemprop' : 'author'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
  
    h=''
    for link in soup.find_all("span", {'itemprop' : 'publisher'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
    
    h=''
    for link in soup.find_all("link", {'itemprop' : 'bookformat'}):
            h=link.get_text()
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')

    
    h=''
    for link in soup.find_all("span", {'itemprop' : 'inLanguage'}):
            h=link.get_text()
            h=h.replace('Language_','')
    if(h!=''):
        l.append(h)
    else:
        l.append('VIDE')
        
        
    k=0
    for link in soup.find("h1"):
        h=''
        if(k/2 == 0):
            l2.append(str(link)[:13])
        else :
            h=str(link)
            h=h.replace('<span itemprop="isbn">','')
            h=h.replace('</span>','')
            l2.append(h)
        k=k+1        
        
    h=''    
    for link in soup.find_all("span", {'class' : 'book-rating-average text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE') 
        
        
    h=''
    for link in soup.find_all("span", {'class' : 'book-rating-provider text-muted'}):
            h=link.get_text()
    if(h!=''):
        l2.append(h)
    else:
        l2.append('VIDE')
        
            
Liste_book.append(book)

            
m=[]
m = [ [l[j], l[j+1], l[j+2],l[j+3],l[j+4]]  for j in range(0, int(len(l)), 5) ]
df_book0 = pandas.DataFrame(index = range(len(m)), columns = ["ISBN","AUTEUR", "PUBLISHER","EDITION","LANGUAGE"])
for i in range(len(m)):
    for variable in range(len(df_book0.columns)):
        df_book0.at[i, df_book0.columns[variable]] = m[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_7_7000.csv"
df_book0.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book0.append(df_book0)

m2=[]
m2 = [ [l2[j], l2[j+1], l2[j+2],l2[j+3],l2[j+4]]  for j in range(0, int(len(l2)), 5) ]
df_book1 = pandas.DataFrame(index = range(len(m2)), columns =["ISBN","ISBN13","ISBN10", "Avg_Rating","Nb_Ratings"])
for i in range(len(m2)):
    for variable in range(len(df_book1.columns)):
        df_book1.at[i, df_book1.columns[variable]] = m2[i][variable]
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_7_7000.csv"
df_book1.to_csv(path, sep=";", encoding="utf-8")
Liste_df_book1.append(df_book1)


In [ ]:
df=df_book1
l=[]



for i in range(df.shape[0]):
    l.append(df["ISBN"][df.index[i]])
    ISBN = str(df["ISBN13"][df.index[i]])
    if(len(ISBN)<13):
        while(len(ISBN)<13):
            ISBN ='0'+ISBN
    url = "https://www.bookdepository.com/book/" + str(ISBN) 
    furl = urllib.request.urlopen(url)
    soup = BeautifulSoup(furl, 'html.parser')
    
    add=''
    for link in soup.find_all('span'):
        if link.get("itemprop")=="numberOfPages":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
    add=''    
    for link in soup.find_all('span'):
        if link.get("itemprop")=="datePublished":
            add=link.get_text()
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        
    add=''
    for link in soup.findAll("div", {'class' : 'item-tools'}):
        for t in link.findAll("span", {'class' : 'sale-price'}):
            add=t.get_text()
    for link in soup.findAll("p", {'class' : 'list-price'}):
            h=link.get_text()
            h=h.replace('\n','')
            h=h.replace('                                            List price: ','')  
            if(add==''):
                add=h
    if(add!=''):
        l.append(add)
    else:
        l.append('VIDE')
        
        

n = [ l[j:j+4]  for j in range(0, int(len(l)), 4) ]
df_book3 = pandas.DataFrame(index = range(len(n)),columns =["ISBN","numberOfPages","datePublished","Price"])
for i in range(len(n)):
    for variable in range(4):
        df_book3.at[i,df_book3.columns[variable]] = n[i][variable]
Liste_df_book3.append(df_book3)

path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_7_7000.csv"
df_book3.to_csv(path, sep=";", encoding="utf-8")



In [ ]:
df_book_Info0_1=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_1_1000.csv",sep=";",index_col=None)
df_book_Info0_2=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_2_2000.csv",sep=";",index_col=None)
df_book_Info0_3=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_3_3000.csv",sep=";",index_col=None)
df_book_Info0_4=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_4_4000.csv",sep=";",index_col=None)
df_book_Info0_5=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_5_5000.csv",sep=";",index_col=None)
df_book_Info0_6=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_6_6000.csv",sep=";",index_col=None)
df_book_Info0_7=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_0_7_7000.csv",sep=";",index_col=None)
frames_Info0 = [df_book_Info0_1, df_book_Info0_2, df_book_Info0_3,df_book_Info0_4,df_book_Info0_5,df_book_Info0_6,df_book_Info0_7]
result_Info0 = pandas.concat(frames_Info0)
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Data_Traitee/result_Info0.csv"
result_Info0.to_csv(path, sep=";", encoding="utf-8")


In [ ]:
df_book_Info1_1=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_1_1000.csv",sep=";",index_col=None)
df_book_Info1_2=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_2_2000.csv",sep=";",index_col=None)
df_book_Info1_3=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_3_3000.csv",sep=";",index_col=None)
df_book_Info1_4=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_4_4000.csv",sep=";",index_col=None)
df_book_Info1_5=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_5_5000.csv",sep=";",index_col=None)
df_book_Info1_6=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_6_6000.csv",sep=";",index_col=None)
df_book_Info1_7=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_1_7_7000.csv",sep=";",index_col=None)
frames_Info1 = [df_book_Info1_1, df_book_Info1_2, df_book_Info1_3,df_book_Info1_4,df_book_Info1_5,df_book_Info1_6,df_book_Info1_7]
result_Info1 = pandas.concat(frames_Info1)
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Data_Traitee/result_Info1.csv"
result_Info1.to_csv(path, sep=";", encoding="utf-8")


In [ ]:
df_book_Info3_1=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_1_1000.csv",sep=";",index_col=None)
df_book_Info3_2=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_2_2000.csv",sep=";",index_col=None)
df_book_Info3_3=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_3_3000.csv",sep=";",index_col=None)
df_book_Info3_4=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_4_4000.csv",sep=";",index_col=None)
df_book_Info3_5=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_5_5000.csv",sep=";",index_col=None)
df_book_Info3_6=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_6_6000.csv",sep=";",index_col=None)
df_book_Info3_7=pandas.DataFrame.from_csv("C:/Users/cooky/Documents/ENSAE_3A/SAND/Book_Info_3_7_7000.csv",sep=";",index_col=None)
frames_Info3 = [df_book_Info3_1, df_book_Info3_2, df_book_Info3_3,df_book_Info3_4,df_book_Info3_5,df_book_Info3_6,df_book_Info3_7]
result_Info3 = pandas.concat(frames_Info3)
path = "C:/Users/cooky/Documents/ENSAE_3A/SAND/Data_Traitee/result_Info3.csv"
result_Info3.to_csv(path, sep=";", encoding="utf-8")
